### PR TITLE
[FLINK-16945][checkpointing] Execute CheckpointFailureManager.FailJob…

### PR DESCRIPTION
**Pull request by ifndef-SleePy**
…Callback directly in main thread executor

## What is the purpose of the change

* This PR is used to resolve the race condition described in [FLINK-13497](https://issues.apache.org/jira/browse/FLINK-13497)
* After finishing [FLINK-13698](https://issues.apache.org/jira/browse/FLINK-13698) , now the operation of failing execution graph should be executed directly in main thread. There will be no unexpected message handled `CheckpointCoordinator` before failing execution graph

## Brief change log

* Invoking `ExecutionGraph#failGlobal` directly in `CheckpointFailureManager`

## Verifying this change

* This change added unit test case

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ExecutionGraphCheckpointCoordinatorTest.java [hunk 1]
```diff
@@ -144,11 +228,14 @@ private ExecutionGraph createExecutionGraphAndEnableCheckpointing(
...
+			.setIoExecutor(Executors.directExecutor())
 			.build();
 
+		final ExecutionJobVertex executionJobVertex = executionGraph.getVerticesTopologically().iterator().next();
+
 		CheckpointCoordinatorConfiguration chkConfig = new CheckpointCoordinatorConfiguration(
-			100,
-			100,
+			1000000,
+			1000000,
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ExecutionGraphCheckpointCoordinatorTest.java:238]: Why did you need to bump those timeouts?

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ExecutionGraphCheckpointCoordinatorTest.java [hunk 2]
```diff
@@ -131,9 +142,82 @@ public void testShutdownCheckpointCoordinatorOnFinished() throws Exception {
...
+		assertThat(checkpointCoordinator, Matchers.notNullValue());
+		assertThat(checkpointCoordinator.isShutdown(), is(false));
+
+		graph.scheduleForExecution();
+		ExecutionGraphTestUtils.switchToRunning(graph);
+
+		// trigger a normal checkpoint which would be declined then
+		checkpointCoordinator.triggerCheckpoint(System.currentTimeMillis(), true);
+		mainThreadExecutor.triggerAll();
+		final long checkpointId = checkpointCoordinator.getPendingCheckpoints().keySet().iterator().next();
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ExecutionGraphCheckpointCoordinatorTest.java:173]: ~Why is this checkpoint going to be declined? Maybe expand the comment with a bit of explanation?~

~If it's failing, maybe add an assertion that it has failed? Or just skip this code block if it's not necessary for the test as a whole?~

edit: 
replace
```
// trigger a normal checkpoint which would be declined then
```
with
```
// trigger a normal checkpoint which we will fail/decline later
```
to avoid confusion as mine above? :)

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ExecutionGraphCheckpointCoordinatorTest.java (hunk 1 above)
Comment (ifndef-SleePy) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ExecutionGraphCheckpointCoordinatorTest.java:238]: Yes, the timeout is one of the reasons. There is another reason that I don't want the periodic checkpoint triggered during testing. It would disturb the verification.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ExecutionGraphCheckpointCoordinatorTest.java (hunk 2 above)
Comment (ifndef-SleePy) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ExecutionGraphCheckpointCoordinatorTest.java]: sure

**Comment by ifndef-SleePy** [comment]
Yes, I think FLINK-13497 is fully resolved after this PR merged.

**Comment by pnowojski** [comment]
Thanks! Merging.
